uc-crux-llvm/src/UCCrux/LLVM/Run/Check.hs

```diff
@@ -0,0 +1,406 @@
+import qualified UCCrux.LLVM.Run.Loop as Loop
+import           UCCrux.LLVM.Run.Result (SomeBugfindingResult)
+import qualified UCCrux.LLVM.Run.Result as Result
+import           UCCrux.LLVM.Setup (SymValue)
+import           UCCrux.LLVM.Shape (Shape)
+{- ORMOLU_ENABLE -}
+
+newtype SomeCheckedCalls m sym arch =
+  SomeCheckedCalls
+    { getSomeCheckedCalls ::
+        forall r.
+        (forall argTypes.
+         DefnSymbol m ->
+         FunctionContext m arch argTypes ->
+         [Check.CheckedCall m sym arch argTypes] ->
+         IO r) ->
+        IO r
+    }
+
+-- | The result of checking inferred contracts
+--
+-- NOTE(lb): The explicit kind signature here is necessary for GHC 8.8/8.6
+-- compatibility.
+newtype CheckResult m arch (argTypes :: Ctx (FullType m)) =
+  CheckResult
+    { getCheckResult ::
+        forall r.
+        (forall sym.
+         IsSymInterface sym =>
+         sym ->
+         -- | Pre-simulation memory
+         MemImpl sym ->
+         -- | Arguments passed to the entry point
+         Assignment (Shape m (SymValue sym arch)) argTypes ->
+         Crux.CruxSimulationResult ->
+         Sim.UCCruxSimulationResult m arch argTypes ->
+         [SomeCheckedCalls m sym arch] ->
+         r) ->
+        r
+    }
+
+data SomeCheckResult m arch =
+  forall argTypes.
+  SomeCheckResult
+    { checkResultFunContext :: FunctionContext m arch argTypes,
+      checkResult :: CheckResult m arch argTypes
+    }
+
+-- NOTE(lb): The explicit kind signature here is necessary for GHC 8.8/8.6
+-- compatibility.
+data TypedConstraints m (argTypes :: Ctx (FullType m))
+  = TypedConstraints
+      { _tcConstraints :: Constraints m argTypes
+      , _tcTypes :: Assignment (FullTypeRepr m) argTypes
+      }
+
+checkInferredContracts ::
+  forall m arch argTypes blocks ret msgs.
+  Crux.Logs msgs =>
+  Crux.SupportsCruxLogMessage msgs =>
+  ArchOk arch =>
+  AppContext ->
+  ModuleContext m arch ->
+  FunctionContext m arch argTypes ->
+  HandleAllocator ->
+  CruxOptions ->
+  LLVMOptions ->
+  Constraints m argTypes ->
+  -- | Entry point
+  Crucible.CFG LLVM blocks (MapToCrucibleType arch argTypes) ret ->
+  -- | Inferred function contracts
+  Map (DefnSymbol m) (Some (TypedConstraints m)) ->
+  IO (CheckResult m arch argTypes)
+checkInferredContracts appCtx modCtx funCtx halloc cruxOpts llOpts constraints cfg contracts =
+   Sim.runSimulatorWithCallbacks
+     appCtx
+     modCtx
+     funCtx
+     halloc
+     constraints
+     cfg
+     cruxOpts
+     llOpts
+     (Sim.SimulatorCallbacks $
+       do ovs <- overrides
+          return $
+            Sim.SimulatorHooks
+              { Sim.createOverrideHooks = map fst ovs
+              , Sim.resultHook =
+                \sym mem args cruxResult ucResult ->
+                  return $
+                    CheckResult $
+                      \k -> k sym mem args cruxResult ucResult (map snd ovs)
+              })
+  where
+    overrides ::
```

Comment:
I'd appreciate a comment here. Is this just creating one checking override for each passed-in contract?

Reply:
Yep, exactly. Comment added.